File: jmccc/src/test/java/org/to2mbn/jmccc/test/ArtifactTest.java
package org.to2mbn.jmccc.test;

import static java.util.Collections.emptyMap;
import static java.util.Collections.singletonMap;
import static org.junit.Assert.assertEquals;
import static org.junit.Assert.assertFalse;
import static org.junit.Assert.assertNull;
import static org.junit.Assert.assertTrue;
import static org.to2mbn.jmccc.version.Artifact.META_SNAPSHOT;
import org.junit.Test;
import org.to2mbn.jmccc.version.Artifact;
import org.to2mbn.jmccc.version.SnapshotVersion;

public class ArtifactTest {

	@Test
	public void test_isSnapshot() {
		assertTrue(new Artifact("group", "artifact", "1.0-SNAPSHOT", "jar", null, emptyMap()).isSnapshot());
	}

	@Test
	public void test_not_isSnapshot() {
		assertFalse(new Artifact("group", "artifact", "1.0", "jar", null, emptyMap()).isSnapshot());
	}

	@Test
	public void test_toString_without_classifier() {
		assertEquals("group:artifact:1.0:jar", new Artifact("group", "artifact", "1.0", "jar", null, emptyMap()).toString());
	}

	@Test
	public void test_toString_with_classifier() {
		assertEquals("group:artifact:1.0:jar:classifier", new Artifact("group", "artifact", "1.0", "jar", "classifier", emptyMap()).toString());
	}

	@Test(expected = IllegalArgumentException.class)
	public void test_resolveSnapshot_failure() {
		new Artifact("group", "artifact", "1.0", "jar", "classifier", emptyMap()).resolveSnapshot(new SnapshotVersion("20170101", 1));
	}

	@Test
	public void test_resolveSnapshot() {
		SnapshotVersion snapshot = new SnapshotVersion("20170101", 1);
		Artifact resolved = new Artifact("group", "artifact", "1.0-SNAPSHOT", "jar", null, emptyMap()).resolveSnapshot(snapshot);
		assertTrue(resolved.isSnapshotResolved());
		assertEquals(snapshot, resolved.getSnapshotVersion());
		assertEquals("1.0-20170101-1", resolved.getResolvedVersion());
	}

	@Test
	public void test_getSnapshotVersion_non_snapshot_version() {
		Artifact artifact = new Artifact("group", "artifact", "1.0", "jar", null, emptyMap());
		assertFalse(artifact.isSnapshotResolved());
		assertNull(artifact.getSnapshotVersion());
		assertEquals("1.0", artifact.getResolvedVersion());
	}

	@Test
	public void test_getSnapshotVersion_non_snapshot_version_with_metadata() {
		SnapshotVersion snapshot = new SnapshotVersion("20170101", 1);
		Artifact artifact = new Artifact("group", "artifact", "1.0", "jar", null, singletonMap(META_SNAPSHOT, snapshot));
		assertFalse(artifact.isSnapshotResolved());
		assertNull(artifact.getSnapshotVersion());
		assertEquals("1.0", artifact.getResolvedVersion());
	}

	@Test
	public void test_getSnapshotVersion_unresolved_snapshot_version() {
		Artifact artifact = new Artifact("group", "artifact", "1.0-SNAPSHOT", "jar", null, emptyMap());
		assertFalse(artifact.isSnapshotResolved());
		assertNull(artifact.getSnapshotVersion());
		assertEquals("1.0-SNAPSHOT", artifact.getResolvedVersion());
	}

}

File: jmccc/src/test/kotlin/org/to2mbn/jmccc/test/ArtifactTest.kt
package org.to2mbn.jmccc.test

import org.junit.Test
import org.to2mbn.jmccc.version.Artifact
import org.to2mbn.jmccc.version.SnapshotVersion
import kotlin.test.assertEquals
import kotlin.test.assertFalse
import kotlin.test.assertNull
import kotlin.test.assertTrue

class ArtifactTest {

	@Test
	fun test_isSnapshot() = assertTrue(
			Artifact(groupId = "group", artifactId = "artifact", version = "1.0-SNAPSHOT").isSnapshot())

	@Test
	fun test_not_isSnapshot() = assertFalse(
			Artifact(groupId = "group", artifactId = "artifact", version = "1.0").isSnapshot())

	@Test
	fun test_toString_without_classifier() = assertEquals(
			"group:artifact:1.0:jar",
			Artifact(groupId = "group", artifactId = "artifact", version = "1.0").toString())

	@Test
	fun test_toString_with_classifier() = assertEquals(
			"group:artifact:1.0:jar:classifier",
			Artifact(groupId = "group", artifactId = "artifact", version = "1.0", classifier = "classifier").toString())

	@Test(expected = IllegalArgumentException::class)
	fun test_resolveSnapshot_failed() {
		Artifact(groupId = "group", artifactId = "artifact", version = "1.0", classifier = "classifier")
				.resolveSnapshot(SnapshotVersion(timestamp = "20170101", buildNumber = 1))
	}

	@Test
	fun test_resolveSnapshot() =
			SnapshotVersion(timestamp = "20170101", buildNumber = 1).let { snapshot ->
				with(Artifact(groupId = "group", artifactId = "artifact", version = "1.0-SNAPSHOT").resolveSnapshot(snapshot)) {
					assertTrue(isSnapshotResolved())
					assertEquals(snapshot, getSnapshotVersion())
					assertEquals("1.0-20170101-1", getResolvedVersion())
				}
			}

	@Test
	fun test_getSnapshotVersion_non_snapshot_version() =
			with(Artifact(groupId = "group", artifactId = "artifact", version = "1.0")) {
				assertFalse(isSnapshotResolved())
				assertNull(getSnapshotVersion())
				assertEquals("1.0", getResolvedVersion())
			}

	@Test
	fun test_getSnapshotVersion_non_snapshot_version_with_metadata() =
			with(Artifact(groupId = "group", artifactId = "artifact", version = "1.0",
					metadata = mapOf(Artifact.META_SNAPSHOT to SnapshotVersion(timestamp = "20170101", buildNumber = 1)))) {
				assertFalse(isSnapshotResolved())
				assertNull(getSnapshotVersion())
				assertEquals("1.0", getResolvedVersion())
			}

	@Test
	fun test_getSnapshotVersion_unresolved_snapshot_version() =
			with(Artifact(groupId = "group", artifactId = "artifact", version = "1.0-SNAPSHOT")) {
				assertFalse(isSnapshotResolved())
				assertNull(getSnapshotVersion())
				assertEquals("1.0-SNAPSHOT", getResolvedVersion())
			}
}
